apps/ensindexer/src/handlers/ThreeDNSToken.ts

```diff
@@ -0,0 +1,302 @@
+import { Context } from "ponder:registry";
+import schema from "ponder:schema";
+import {
+  makeSharedEventValues,
+  upsertAccount,
+  upsertDomain,
+  upsertRegistration,
+} from "@/lib/db-helpers";
+import { labelByLabelHash } from "@/lib/graphnode-helpers";
+import { makeRegistrationId } from "@/lib/ids";
+import { parseLabelAndNameFromOnChainMetadata } from "@/lib/plugin-helpers";
+import { EventWithArgs } from "@/lib/ponder-helpers";
+import { encodeLabelhash } from "@ensdomains/ensjs/utils";
+import { LabelHash, Node, PluginName } from "@ensnode/utils";
+import {
+  decodeDNSPacketBytes,
+  isLabelIndexable,
+  makeSubdomainNode,
+} from "@ensnode/utils/subname-helpers";
+import { Address, Hex, hexToBigInt, hexToBytes, labelhash } from "viem";
+
+/**
+ * makes a set of shared handlers for a ThreeDNSToken contract
+ */
+export const makeThreeDNSTokenHandlers = ({
+  pluginName,
+  getUriForTokenId,
+}: {
+  pluginName: PluginName;
+  getUriForTokenId: (context: Context, tokenId: bigint) => Promise<string>;
+}) => {
+  const sharedEventValues = makeSharedEventValues(pluginName);
+
+  return {
+    /**
+     * In ThreeDNS, NewOwner is emitted when a (sub)domain is created. This includes TLDs, 2LDs, and
+     * >2LDs. For 2LDs, however, RegistrationCreated is always emitted, and it's emitted first, so
+     * this function must upsert a Domain that may have been created in `handleRegistrationCreated`.
+     *
+     * Finally, NewOwner can be emitted for the same Domain across chains — ThreeDNS allows registrations of
+     * .xyz TLDs, for example, on both Optimism and Base, so this function must be idempotent along
+     * that dimension as well.
+     */
+    handleNewOwner: async ({
+      context,
+      event,
+    }: {
+      context: Context;
+      event: EventWithArgs<{
+        // NOTE: `node` event arg represents a `Node` that is the _parent_ of the node the NewOwner event is about
+        node: Node;
+        // NOTE: `label` event arg represents a `LabelHash` for the sub-node under `node`
+        label: LabelHash;
+        owner: Address;
+      }>;
+    }) => {
+      const { label: labelHash, node: parentNode, owner } = event.args;
+
+      await upsertAccount(context, owner);
+
+      // the domain in question is a subdomain of `parentNode`
+      const node = makeSubdomainNode(labelHash, parentNode);
+
+      const domain = await upsertDomain(context, {
+        id: node,
+        ownerId: owner,
+        parentId: parentNode,
+        createdAt: event.block.timestamp,
+        labelhash: labelHash,
+
+        // NOTE: threedns has no concept of registry migration, so domains managed by this plugin
+        // are always considered migrated
+        isMigrated: true,
+      });
```

Comment:
Noting that we didn't include `subdomainCount` value updates.

Reply:
implemented, thanks for the heads up